streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java

```diff
@@ -611,65 +628,170 @@ public StateStore getStateStore(final String name) {
+     * interface. <strong>It is highly recommended to update the code for this case to avoid bugs and to use
+     * {@link #getTimestampedKeyValueStore(String)} for full store access instead.</strong>
+     * <p>
      * This is often useful in test cases to pre-populate the store before the test case instructs the topology to
      * {@link #pipeInput(ConsumerRecord) process an input message}, and/or to check the store afterward.
      *
      * @param name the name of the store
-     * @return the key value store, or {@code null} if no {@link KeyValueStore} has been registered with the given name
+     * @return the key value store, or {@code null} if no {@link KeyValueStore} or {@link TimestampedKeyValueStore}
+     * has been registered with the given name
      * @see #getAllStateStores()
      * @see #getStateStore(String)
+     * @see #getTimestampedKeyValueStore(String)
      * @see #getWindowStore(String)
+     * @see #getTimestampedWindowStore(String)
      * @see #getSessionStore(String)
+     * @see #getTimestampedSessionStore(String)
      */
     @SuppressWarnings({"unchecked", "WeakerAccess"})
     public <K, V> KeyValueStore<K, V> getKeyValueStore(final String name) {
         final StateStore store = getStateStore(name);
+        if (store instanceof TimestampedKeyValueStore) {
+            log.warn("Method #getTimestampedKeyValueStore() should be used to access a TimestampedKeyValueStore.");
+            return new KeyValueStoreFacade<>((TimestampedKeyValueStore<K, V>) store);
+        }
         return store instanceof KeyValueStore ? (KeyValueStore<K, V>) store : null;
     }
 
     /**
-     * Get the {@link WindowStore} with the given name.
+     * Get the {@link TimestampedKeyValueStore} with the given name.
      * The store can be a "regular" or global store.
      * <p>
      * This is often useful in test cases to pre-populate the store before the test case instructs the topology to
      * {@link #pipeInput(ConsumerRecord) process an input message}, and/or to check the store afterward.
      *
      * @param name the name of the store
-     * @return the key value store, or {@code null} if no {@link WindowStore} has been registered with the given name
+     * @return the key value store, or {@code null} if no {@link TimestampedKeyValueStore} has been registered with the given name
      * @see #getAllStateStores()
      * @see #getStateStore(String)
      * @see #getKeyValueStore(String)
-     * @see #getSessionStore(String) (String)
+     * @see #getWindowStore(String)
+     * @see #getTimestampedWindowStore(String)
+     * @see #getSessionStore(String)
+     * @see #getTimestampedSessionStore(String)
      */
-    @SuppressWarnings({"unchecked", "WeakerAccess", "unused"})
+    @SuppressWarnings({"unchecked", "WeakerAccess"})
+    public <K, V> TimestampedKeyValueStore<K, V> getTimestampedKeyValueStore(final String name) {
```

Comment:
It seems like a minor discrepancy with the `KafkaStreams#store` (IQ) counterpart... The IQ method will give back a `KeyValueStore<K, ValueAndTimestamp<V>>`, but this gives back a `TimestampedKeyValueStore<K,V>`.

I think I like the former one better, since it is more generic.

Can we change the return types of the TestDriver methods to match the IQ returned store types? It's an update to the kip, but it seems like a trivially minor one.

Reply:
I see your point. However, for IQ we return read-only stores (in contrast to the test driver).

To avoid adding new interfaces, we simply return `ReadOnlyXxxStore<K, ValueAndTimestamp<V>>` in IQ. We would need to introduce `ReadOnlyTimestampedXxxStore<K, V>` interfaces otherwise.

For the test driver, it will if fact return a `TimestampedKeyValueStore<K, V>` though that can be casted to `KeyValueStore<K, ValueAndTimestamp<V>>`. We needed to return `TimestampedKeyValueStore` originally, because we added new methods (that are now removed).

I personally don't see a big difference between both. Because the return type in IQ is different anyway (read-only vs read-write). I also don't see a big advantage in "aligning" the return types -- as I don't see an "alignment" to begin with (it's not the same type). (If we want the alignment, we should not introduce new interface, but follow your suggestion.)

If @guozhangwang @bbejeck think we should change it to `KeyValueStore<K, ValueAndTimestamp<V>>`, I am open to do the change. Just don't see a big necessity from my personal point of view.

Reply:
@mjsax , Thanks for that explanation. That makes sense why IQ needs to return the more generic type.

I'd say I'm +0.7 for changing to the return type here to `KeyValueStore<K, ValueAndTimestamp<V>>`.
1. The type is more descriptive for users about what they can expect from the store
2. The marker interface only exists for internal handling of the store. It's conflating the purpose of the interface to use it both as a marker interface and as an actual type.
3. No cast is necessary, since `KeyValueStore<K, ValueAndTimestamp<V>>` is a supertype of `TimestampedKeyValueStore<K, V>`. When we get the `TimestampedKeyValueStore<K, V>`, we can simply return/assign it as a `KeyValueStore<K, ValueAndTimestamp<V>>`.
4. (a little bit repetitive from point 2) There's no value to the TTD users in giving them an extra interface to handle. More interfaces means more API surface area, which incrementally increases the mental load of using Streams.

At the end of the day, I don't think it's *super* harmful, so if you or the others prefer to keep it as-is, it's fine. Since this is a public API, I think it pays to consider it now, though.

Reply:
Initially, I was in favor of aligning the return types, but now that I remember we return `ReadOnlyxxx` variants we can't truly align without additional changes and considering I'm not sure it buys us much I'm now in favor of leaving as is.

Reply:
It's not so much about making the store the same as IQ as it is about not using the marker interface as *both* a marker *and* a caller-facing type. IMHO, there's no reason to do it, and no actual benefit from doing it, but there are downsides...

Reply:
I'd support @vvcephei 's argument to be consistent with IQ.

Reply:
Ack